lib/landing_session.js

```diff
@@ -65,34 +68,44 @@ class LandingSession extends Session {
   }
 
   async downloadAndPatch() {
-    const { cli, req, repo, owner, prid } = this;
+    const { cli, repo, owner, prid, expectedCommitShas } = this;
 
-    // TODO(joyeecheung): restore previously downloaded patches
     cli.startSpinner(`Downloading patch for ${prid}`);
-    const patch = await req.text(
-      `https://github.com/${owner}/${repo}/pull/${prid}.patch`);
-    this.savePatch(patch);
-    cli.stopSpinner(`Downloaded patch to ${this.patchPath}`);
+    await runAsync('git', [
+      'fetch', `https://github.com/${owner}/${repo}.git`,
+      `refs/pull/${prid}/merge`]);
+    const [base, head] = (await runAsync('git', [
+      'rev-parse', 'FETCH_HEAD^1', 'FETCH_HEAD^2'], { captureStdout: true }))
+      .split('\n');
+    const commitShas = (await runAsync('git', [
+      'rev-list', `${base}..${head}`], { captureStdout: true }))
+      .trim().split('\n');
+    cli.stopSpinner(`Fetched commits as ${shortSha(base)}..${shortSha(head)}`);
     cli.separator();
-    // TODO: check that patches downloaded match metadata.commits
+
+    const mismatchedCommits = [
+      ...commitShas.filter((sha) => !expectedCommitShas.includes(sha))
+        .map((sha) => `Unexpected commit ${sha}`),
+      ...expectedCommitShas.filter((sha) => !commitShas.includes(sha))
+        .map((sha) => `Missing commit ${sha}`)
+    ].join('\n');
+    if (mismatchedCommits.length > 0) {
+      cli.error(`Mismatched commits:\n${mismatchedCommits}`);
+      process.exit(1);
+    }
+
+    const commitInfo = { base, head, shas: commitShas };
+    this.saveCommitInfo(commitInfo);
+
     try {
-      await forceRunAsync('git', ['am', this.patchPath], {
+      await forceRunAsync('git', ['cherry-pick', `${base}..${head}`], {
```

Comment:
Can't we `git checkout head && git rebase` instead? What's the advantage of using cherry-pick? With rebase we could autosquash and amend by default, which would considerably simplify the workflow.

(we can do this in a future PR too, not blocking)

Reply:
I think so – this was just closer to the original concept of using `git am` to apply the patch

If we do want to autosquash here, that would change the structure a lot more, right? We’d essentially be merging two steps into one?

Reply:
> If we do want to autosquash here, that would change the structure a lot more, right? We’d essentially be merging two steps into one?

Right, which is why I think it would simplify the workflow. But definitely doesn't need to be part of this PR :)